File: javaspider/src/main/java/com/javaspider/media/DASHDownloader.java
package com.javaspider.media;

import org.w3c.dom.*;
import javax.xml.parsers.*;
import java.io.*;
import java.net.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;
import java.util.regex.*;

/**
 * DASH 流媒体下载器
 * 支持 MPD 解析和多清晰度选择
 */
public class DASHDownloader {
    
    private String outputDir;
    private int concurrent;
    private int timeout;
    private int retryTimes;
    private String userAgent;
    private String referer;
    private volatile boolean cancelled;
    
    /**
     * 默认构造函数
     */
    public DASHDownloader() {
        this("./downloads");
    }
    
    /**
     * 指定输出目录
     * @param outputDir 输出目录
     */
    public DASHDownloader(String outputDir) {
        this.outputDir = outputDir;
        this.concurrent = 5;
        this.timeout = 30000;
        this.retryTimes = 3;
        this.userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";
        this.cancelled = false;
    }
    
    /**
     * 设置并发数
     */
    public void setConcurrent(int concurrent) {
        this.concurrent = concurrent;
    }
    
    /**
     * 设置超时
     */
    public void setTimeout(int timeout) {
        this.timeout = timeout;
    }
    
    /**
     * 设置重试次数
     */
    public void setRetryTimes(int retryTimes) {
        this.retryTimes = retryTimes;
    }
    
    /**
     * 设置 User-Agent
     */
    public void setUserAgent(String userAgent) {
        this.userAgent = userAgent;
    }
    
    /**
     * 设置 Referer
     */
    public void setReferer(String referer) {
        this.referer = referer;
    }
    
    /**
     * 取消下载
     */
    public void cancel() {
        this.cancelled = true;
    }
    
    /**
     * 下载 DASH 流
     * @param mpdUrl MPD URL
     * @param quality 清晰度 (4k, 2k, 1080p, 720p, auto)
     * @return 下载结果
     */
    public DownloadResult download(String mpdUrl, String quality) {
        long startTime = System.currentTimeMillis();
        
        try {
            // 创建输出目录
            Files.createDirectories(Paths.get(outputDir));
            
            // 下载并解析 MPD
            MPDInfo mpdInfo = parseMPD(mpdUrl);
            if (mpdInfo == null) {
                return new DownloadResult(false, "解析 MPD 失败");
            }
            
            // 选择最佳格式
            FormatInfo videoFormat = selectVideoFormat(mpdInfo.videoFormats, quality);
            FormatInfo audioFormat = selectAudioFormat(mpdInfo.audioFormats);
            
            if (videoFormat == null) {
                return new DownloadResult(false, "未找到视频格式");
            }
            
            System.out.println("选择视频格式：" + videoFormat.quality + " (" + 
                videoFormat.width + "x" + videoFormat.height + ")");
            
            if (audioFormat != null) {
                System.out.println("选择音频格式：" + audioFormat.codecs);
            }
            
            // 下载目录
            String downloadDir = outputDir + "/dash_temp_" + System.currentTimeMillis();
            Files.createDirectories(Paths.get(downloadDir));
            
            // 下载视频
            String videoFile = null;
            if (videoFormat != null) {
                System.out.println("正在下载视频分段...");
                videoFile = downloadDir + "/video.mp4";
                List<String> videoFiles = downloadSegments(videoFormat.urls, downloadDir, "video");
                if (videoFiles == null || videoFiles.isEmpty()) {
                    return new DownloadResult(false, "视频下载失败");
                }
                mergeFiles(videoFiles, videoFile);
            }
            
            // 下载音频
            String audioFile = null;
            if (audioFormat != null) {
                System.out.println("正在下载音频分段...");
                audioFile = downloadDir + "/audio.mp4";
                List<String> audioFiles = downloadSegments(audioFormat.urls, downloadDir, "audio");
                if (audioFiles != null && !audioFiles.isEmpty()) {
                    mergeFiles(audioFiles, audioFile);
                }
            }
            
            // 合并音视频
            String outputFile = outputDir + "/output_" + System.currentTimeMillis() + ".mp4";
            if (videoFile != null && audioFile != null) {
                // 有 ffmpeg 则合并，没有则只保存视频
                outputFile = mergeAudioVideo(videoFile, audioFile, outputFile);
            } else if (videoFile != null) {
                outputFile = videoFile;
            } else if (audioFile != null) {
                outputFile = audioFile;
            }
            
            // 清理临时目录
            cleanupTempDir(downloadDir);
            
            long duration = System.currentTimeMillis() - startTime;
            return new DownloadResult(true, "下载完成", outputFile, duration);
            
        } catch (Exception e) {
            e.printStackTrace();
            return new DownloadResult(false, "下载失败：" + e.getMessage());
        }
    }
    
    /**
     * 下载并解析 MPD
     */
    private MPDInfo parseMPD(String mpdUrl) throws Exception {
        String content = downloadContent(mpdUrl);
        if (content == null) {
            return null;
        }
        
        // 解析 XML
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document doc = builder.parse(new ByteArrayInputStream(content.getBytes()));
        
        MPDInfo info = new MPDInfo();
        String basePath = mpdUrl.substring(0, mpdUrl.lastIndexOf('/') + 1);
        
        // 提取 AdaptationSet
        NodeList adaptationSets = doc.getElementsByTagName("AdaptationSet");
        
        for (int i = 0; i < adaptationSets.getLength(); i++) {
            Element adaptationSet = (Element) adaptationSets.item(i);
            String mimeType = adaptationSet.getAttribute("mimeType");
            
            NodeList representations = adaptationSet.getElementsByTagName("Representation");
            
            for (int j = 0; j < representations.getLength(); j++) {
                Element rep = (Element) representations.item(j);
                
                FormatInfo format = new FormatInfo();
                format.id = rep.getAttribute("id");
                format.mimeType = mimeType;
                
                String width = rep.getAttribute("width");
                String height = rep.getAttribute("height");
                String bandwidth = rep.getAttribute("bandwidth");
                String codecs = rep.getAttribute("codecs");
                
                format.width = width.isEmpty() ? 0 : Integer.parseInt(width);
                format.height = height.isEmpty() ? 0 : Integer.parseInt(height);
                format.bandwidth = bandwidth.isEmpty() ? 0 : Integer.parseInt(bandwidth);
                format.codecs = codecs;
                
                // 确定质量
                format.quality = determineQuality(format.height);
                
                // 确定类型
                format.isVideo = mimeType.contains("video") || format.width > 0;
                format.isAudio = mimeType.contains("audio") || codecs.contains("mp4a");
                
                // 提取分段 URL
                List<String> urls = extractSegmentUrls(rep, basePath);
                format.urls = urls;
                
                if (!urls.isEmpty()) {
                    if (format.isVideo) {
                        info.videoFormats.add(format);
                    } else if (format.isAudio) {
                        info.audioFormats.add(format);
                    }
                }
            }
        }
        
        return info;
    }
    
    /**
     * 提取分段 URL
     */
    private List<String> extractSegmentUrls(Element rep, String basePath) {
        List<String> urls = new ArrayList<>();
        
        // SegmentList
        NodeList segmentLists = rep.getElementsByTagName("SegmentList");
        if (segmentLists.getLength() > 0) {
            Element segmentList = (Element) segmentLists.item(0);
            NodeList segmentUrls = segmentList.getElementsByTagName("SegmentURL");
            
            for (int i = 0; i < segmentUrls.getLength(); i++) {
                Element segUrl = (Element) segmentUrls.item(i);
                String media = segUrl.getAttribute("media");
                if (!media.isEmpty()) {
                    urls.add(absoluteUrl(media, basePath));
                }
            }
        }
        
        // SegmentTemplate
        NodeList segmentTemplates = rep.getElementsByTagName("SegmentTemplate");
        if (segmentTemplates.getLength() > 0) {
            Element segmentTemplate = (Element) segmentTemplates.item(0);
            String media = segmentTemplate.getAttribute("media");
            String startNumber = segmentTemplate.getAttribute("startNumber");
            String duration = segmentTemplate.getAttribute("duration");
            String timescale = segmentTemplate.getAttribute("timescale");
            
            if (!media.isEmpty()) {
                int start = startNumber.isEmpty() ? 1 : Integer.parseInt(startNumber);
                int count = 100; // 默认分段数
                
                for (int i = 0; i < count; i++) {
                    String url = media
                        .replace("$Number$", String.format("%04d", start + i))
                        .replace("$RepresentationID$", rep.getAttribute("id"));
                    urls.add(absoluteUrl(url, basePath));
                }
            }
        }
        
        return urls;
    }
    
    /**
     * 下载内容
     */
    private String downloadContent(String url) throws IOException {
        for (int i = 0; i < retryTimes; i++) {
            try {
                HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
                conn.setRequestProperty("User-Agent", userAgent);
                if (referer != null && !referer.isEmpty()) {
                    conn.setRequestProperty("Referer", referer);
                }
                conn.setConnectTimeout(timeout);
                conn.setReadTimeout(timeout);
                
                if (conn.getResponseCode() != 200) {
                    conn.disconnect();
                    continue;
                }
                
                BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
                StringBuilder content = new StringBuilder();
                String line;
                
                while ((line = reader.readLine()) != null) {
                    content.append(line).append("\n");
                }
                
                reader.close();
                conn.disconnect();
                return content.toString();
                
            } catch (Exception e) {
                if (i == retryTimes - 1) {
                    throw new IOException("下载失败：" + e.getMessage());
                }
            }
        }
        return null;
    }
    
    /**
     * 下载分段
     */
    private List<String> downloadSegments(List<String> urls, String dir, String prefix) {
        List<String> files = new CopyOnWriteArrayList<>();
        ExecutorService executor = Executors.newFixedThreadPool(concurrent);
        CountDownLatch latch = new CountDownLatch(urls.size());
        AtomicInteger successCount = new AtomicInteger(0);
        
        for (int i = 0; i < urls.size(); i++) {
            final int index = i;
            final String url = urls.get(i);
            
            executor.submit(() -> {
                try {
                    if (cancelled) return;
                    
                    String filename = String.format("%s_%04d.tmp", prefix, index);
                    String filepath = dir + "/" + filename;
                    
                    if (downloadFile(url, filepath)) {
                        files.add(filepath);
                        successCount.incrementAndGet();
                        System.out.printf("\r进度：%d/%d", index + 1, urls.size());
                    }
                    
                } finally {
                    latch.countDown();
                }
            });
        }
        
        try {
            latch.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        
        executor.shutdown();
        System.out.println();
        
        return files;
    }
    
    /**
     * 下载文件
     */
    private boolean downloadFile(String url, String filepath) {
        for (int i = 0; i < retryTimes; i++) {
            try {
                HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
                conn.setRequestProperty("User-Agent", userAgent);
                conn.setConnectTimeout(timeout);
                conn.setReadTimeout(timeout);
                
                if (conn.getResponseCode() != 200) {
                    conn.disconnect();
                    continue;
                }
                
                try (InputStream in = conn.getInputStream();
                     FileOutputStream out = new FileOutputStream(filepath)) {
                    
                    byte[] buffer = new byte[8192];
                    int len;
                    
                    while ((len = in.read(buffer)) != -1) {
                        out.write(buffer, 0, len);
                    }
                }
                
                conn.disconnect();
                return true;
                
            } catch (Exception e) {
                if (i == retryTimes - 1) {
                    return false;
                }
            }
        }
        return false;
    }
    
    /**
     * 合并文件
     */
    private void mergeFiles(List<String> files, String outputFile) throws IOException {
        Collections.sort(files);
        
        try (FileOutputStream out = new FileOutputStream(outputFile)) {
            for (String file : files) {
                byte[] data = Files.readAllBytes(Paths.get(file));
                out.write(data);
            }
        }
    }
    
    /**
     * 合并音视频
     */
    private String mergeAudioVideo(String videoFile, String audioFile, String outputFile) {
        // 尝试使用 FFmpeg
        String ffmpegPath = findFFmpeg();
        if (ffmpegPath != null) {
            try {
                ProcessBuilder pb = new ProcessBuilder(
                    ffmpegPath,
                    "-i", videoFile,
                    "-i", audioFile,
                    "-c:v", "copy",
                    "-c:a", "aac",
                    "-y",
                    outputFile
                );
                pb.inheritIO();
                Process process = pb.start();
                process.waitFor();
                
                if (process.exitValue() == 0) {
                    System.out.println("FFmpeg 合并成功");
                    return outputFile;
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        
        // 没有 FFmpeg，只返回视频
        System.out.println("未找到 FFmpeg，仅保存视频");
        return videoFile;
    }
    
    /**
     * 查找 FFmpeg
     */
    private String findFFmpeg() {
        // 检查环境变量
        String ffmpegPath = System.getenv("FFMPEG_PATH");
        if (ffmpegPath != null && !ffmpegPath.isEmpty()) {
            return ffmpegPath;
        }
        
        // 检查 PATH
        String[] paths = {"ffmpeg", "C:\\ffmpeg\\bin\\ffmpeg.exe", "/usr/bin/ffmpeg", "/usr/local/bin/ffmpeg"};
        for (String path : paths) {
            try {
                ProcessBuilder pb = new ProcessBuilder(path, "-version");
                Process process = pb.start();
                process.waitFor();
                if (process.exitValue() == 0) {
                    return path;
                }
            } catch (Exception e) {
                // 忽略
            }
        }
        
        return null;
    }
    
    /**
     * 清理临时目录
     */
    private void cleanupTempDir(String dir) {
        try {
            Files.walk(Paths.get(dir))
                .sorted(Comparator.reverseOrder())
                .forEach(path -> {
                    try {
                        Files.delete(path);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                });
            Files.delete(Paths.get(dir));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    
    /**
     * 绝对 URL
     */
    private String absoluteUrl(String url, String basePath) {
        if (url.startsWith("http://") || url.startsWith("https://")) {
            return url;
        }
        if (url.startsWith("/")) {
            try {
                URL base = new URL(basePath);
                return base.getProtocol() + "://" + base.getHost() + url;
            } catch (Exception e) {
                return basePath + url;
            }
        }
        return basePath + url;
    }
    
    /**
     * 确定质量
     */
    private String determineQuality(int height) {
        if (height >= 2160) return "4K";
        if (height >= 1440) return "2K";
        if (height >= 1080) return "1080p";
        if (height >= 720) return "720p";
        if (height >= 480) return "480p";
        return "360p";
    }
    
    /**
     * 选择视频格式
     */
    private FormatInfo selectVideoFormat(List<FormatInfo> formats, String quality) {
        if (formats.isEmpty()) return null;
        
        if ("auto".equals(quality)) {
            // 选择最高清晰度
            FormatInfo best = formats.get(0);
            for (FormatInfo f : formats) {
                if (f.height > best.height) {
                    best = f;
                }
            }
            return best;
        }
        
        Map<String, Integer> qualityMap = new HashMap<>();
        qualityMap.put("4k", 2160);
        qualityMap.put("2k", 1440);
        qualityMap.put("1080p", 1080);
        qualityMap.put("720p", 720);
        qualityMap.put("480p", 480);
        qualityMap.put("360p", 360);
        
        int targetHeight = qualityMap.getOrDefault(quality.toLowerCase(), 1080);
        
        FormatInfo best = null;
        for (FormatInfo f : formats) {
            if (f.height <= targetHeight) {
                if (best == null || f.height > best.height) {
                    best = f;
                }
            }
        }
        
        return best != null ? best : formats.get(0);
    }
    
    /**
     * 选择音频格式
     */
    private FormatInfo selectAudioFormat(List<FormatInfo> formats) {
        if (formats.isEmpty()) return null;
        
        FormatInfo best = formats.get(0);
        for (FormatInfo f : formats) {
            if (f.bandwidth > best.bandwidth) {
                best = f;
            }
        }
        return best;
    }
    
    /**
     * MPD 信息
     */
    static class MPDInfo {
        List<FormatInfo> videoFormats = new ArrayList<>();
        List<FormatInfo> audioFormats = new ArrayList<>();
    }
    
    /**
     * 格式信息
     */
    static class FormatInfo {
        String id;
        String mimeType;
        String quality;
        int width;
        int height;
        int bandwidth;
        String codecs;
        boolean isVideo;
        boolean isAudio;
        List<String> urls = new ArrayList<>();
    }
    
    /**
     * 下载结果
     */
    public static class DownloadResult {
        public boolean success;
        public String message;
        public String outputFile;
        public long duration;
        
        public DownloadResult(boolean success, String message) {
            this.success = success;
            this.message = message;
        }
        
        public DownloadResult(boolean success, String message, String outputFile, long duration) {
            this.success = success;
            this.message = message;
            this.outputFile = outputFile;
            this.duration = duration;
        }
    }
}

File: gospider/media/dash_downloader.go
package media

import (
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// LegacyDASHDownloader 保留旧版 DASH 下载器实现，避免与增强版 API 冲突。
type LegacyDASHDownloader struct {
	client     *http.Client
	outputDir  string
	userAgent  string
	concurrent int
}

// MPD MPD 播放列表
type MPD struct {
	XMLName     xml.Name `xml:"MPD"`
	Type        string   `xml:"type,attr"`
	MediaPres   string   `xml:"mediaPresentationDuration,attr"`
	Periods     []Period `xml:"Period"`
}

// Period 周期
type Period struct {
	Duration     string      `xml:"duration,attr"`
	AdaptationSets []AdaptationSet `xml:"AdaptationSet"`
}

// AdaptationSet 自适应集
type AdaptationSet struct {
	MimeType    string     `xml:"mimeType,attr"`
	Representations []Representation `xml:"Representation"`
}

// Representation 表示
type Representation struct {
	ID          string   `xml:"id,attr"`
	Width       int      `xml:"width,attr"`
	Height      int      `xml:"height,attr"`
	Bandwidth   int      `xml:"bandwidth,attr"`
	Codecs      string   `xml:"codecs,attr"`
	SegmentBase *SegmentBase `xml:"SegmentBase"`
	SegmentList *SegmentList `xml:"SegmentList"`
	SegmentTemplate *SegmentTemplate `xml:"SegmentTemplate"`
}

// SegmentBase 分段基础
type SegmentBase struct {
	Initialization *Initialization `xml:"Initialization"`
}

// SegmentList 分段列表
type SegmentList struct {
	Initialization *Initialization `xml:"Initialization"`
	SegmentURLs    []SegmentURL    `xml:"SegmentURL"`
}

// SegmentTemplate 分段模板
type SegmentTemplate struct {
	Media      string `xml:"media,attr"`
	Initialization string `xml:"initialization,attr"`
	StartNumber int    `xml:"startNumber,attr"`
	Timescale  int    `xml:"timescale,attr"`
	Duration   int    `xml:"duration,attr"`
}

// Initialization 初始化分段
type Initialization struct {
	SourceURL string `xml:"sourceURL,attr"`
	Range     string `xml:"range,attr"`
}

// SegmentURL 分段 URL
type SegmentURL struct {
	Media      string `xml:"media,attr"`
	MediaRange string `xml:"mediaRange,attr"`
}

// LegacyDASHFormat 是旧版下载器使用的格式表示。
type LegacyDASHFormat struct {
	ID        string
	Quality   string
	Width     int
	Height    int
	Bandwidth int
	Codecs    string
	URLs      []string
}

// NewLegacyDASHDownloader 创建旧版 DASH 下载器。
func NewLegacyDASHDownloader(outputDir string) *LegacyDASHDownloader {
	return &LegacyDASHDownloader{
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
		outputDir:  outputDir,
		userAgent:  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
		concurrent: 5,
	}
}

// DownloadMPD 下载 MPD 文件
func (d *LegacyDASHDownloader) DownloadMPD(mpdURL string) (*MPD, error) {
	req, err := http.NewRequest("GET", mpdURL, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", d.userAgent)

	resp, err := d.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var mpd MPD
	if err := xml.Unmarshal(body, &mpd); err != nil {
		return nil, err
	}

	return &mpd, nil
}

// ParseMPD 解析 MPD 提取格式信息
func (d *LegacyDASHDownloader) ParseMPD(mpd *MPD, baseURL string) []LegacyDASHFormat {
	formats := make([]LegacyDASHFormat, 0)

	for _, period := range mpd.Periods {
		for _, adaptation := range period.AdaptationSets {
			for _, rep := range adaptation.Representations {
				format := LegacyDASHFormat{
					ID:        rep.ID,
					Width:     rep.Width,
					Height:    rep.Height,
					Bandwidth: rep.Bandwidth,
					Codecs:    rep.Codecs,
					URLs:      make([]string, 0),
				}

				// 确定质量
				if rep.Height >= 2160 {
					format.Quality = "4K"
				} else if rep.Height >= 1440 {
					format.Quality = "2K"
				} else if rep.Height >= 1080 {
					format.Quality = "1080p"
				} else if rep.Height >= 720 {
					format.Quality = "720p"
				} else if rep.Height >= 480 {
					format.Quality = "480p"
				} else {
					format.Quality = "360p"
				}

				// 提取分段 URL
				if rep.SegmentList != nil {
					for _, seg := range rep.SegmentList.SegmentURLs {
						url := seg.Media
						if !strings.HasPrefix(url, "http") {
							url = baseURL + url
						}
						format.URLs = append(format.URLs, url)
					}
				}

				// 处理 SegmentTemplate
				if rep.SegmentTemplate != nil {
					// 动态 MPD 需要特殊处理
					fmt.Printf("注意：SegmentTemplate 需要动态生成 URL\n")
				}

				if len(format.URLs) > 0 {
					formats = append(formats, format)
				}
			}
		}
	}

	return formats
}

// DownloadDASH 下载 DASH 流
func (d *LegacyDASHDownloader) DownloadDASH(mpdURL, outputFile string) error {
	// 下载 MPD
	mpd, err := d.DownloadMPD(mpdURL)
	if err != nil {
		return fmt.Errorf("下载 MPD 失败：%v", err)
	}

	// 解析 MPD
	baseURL := mpdURL[:strings.LastIndex(mpdURL, "/")+1]
	formats := d.ParseMPD(mpd, baseURL)

	if len(formats) == 0 {
		return fmt.Errorf("未找到任何媒体流")
	}

	// 选择最佳视频和音频格式
	var bestVideo, bestAudio LegacyDASHFormat
	for _, format := range formats {
		if strings.Contains(format.Codecs, "avc") || strings.Contains(format.Codecs, "hevc") {
			if bestVideo.ID == "" || format.Bandwidth > bestVideo.Bandwidth {
				bestVideo = format
			}
		}
		if strings.Contains(format.Codecs, "mp4a") || strings.Contains(format.Codecs, "ac-3") {
			if bestAudio.ID == "" || format.Bandwidth > bestAudio.Bandwidth {
				bestAudio = format
			}
		}
	}

	fmt.Printf("选择视频格式：%s (%dx%d, %d kbps)\n", 
		bestVideo.Quality, bestVideo.Width, bestVideo.Height, bestVideo.Bandwidth/1000)
	
	if bestAudio.ID != "" {
		fmt.Printf("选择音频格式：%s (%d kbps)\n", bestAudio.Quality, bestAudio.Bandwidth/1000)
	}

	// 创建临时目录
	tempDir := filepath.Join(d.outputDir, "dash_temp")
	os.MkdirAll(tempDir, 0755)
	defer os.RemoveAll(tempDir)

	// 下载视频分段
	fmt.Println("正在下载视频分段...")
	videoFile := filepath.Join(tempDir, "video.mp4")
	if err := d.downloadSegments(bestVideo.URLs, videoFile); err != nil {
		return fmt.Errorf("下载视频失败：%v", err)
	}

	// 下载音频分段（如果有）
	audioFile := ""
	if bestAudio.ID != "" {
		fmt.Println("正在下载音频分段...")
		audioFile = filepath.Join(tempDir, "audio.mp4")
		if err := d.downloadSegments(bestAudio.URLs, audioFile); err != nil {
			return fmt.Errorf("下载音频失败：%v", err)
		}
	}

	// 合并音视频
	fmt.Println("正在合并音视频...")
	if audioFile != "" {
		ffmpeg := NewFFmpegWrapper("", d.outputDir)
		if ffmpegPath, err := AutoDetectFFmpeg(); err == nil {
			ffmpeg.FFmpegPath = ffmpegPath
			return ffmpeg.CombineAudioVideo(videoFile, audioFile, outputFile)
		}
		// 如果没有 ffmpeg，只保存视频
		os.Rename(videoFile, outputFile)
	} else {
		os.Rename(videoFile, outputFile)
	}

	return nil
}

// downloadSegments 下载分段
func (d *LegacyDASHDownloader) downloadSegments(urls []string, outputFile string) error {
	file, err := os.Create(outputFile)
	if err != nil {
		return err
	}
	defer file.Close()

	downloader := NewEnhancedDownloader(d.outputDir)
	downloader.SetConcurrent(d.concurrent)

	for i, url := range urls {
		tempFile := filepath.Join(d.outputDir, fmt.Sprintf("seg_%d.tmp", i))
		
		_, err := downloader.DownloadWithProgress(url, tempFile)
		if err != nil {
			return err
		}

		// 读取并写入
		data, err := os.ReadFile(tempFile)
		if err != nil {
			return err
		}
		file.Write(data)
		
		os.Remove(tempFile)
		
		fmt.Printf("\r进度：%d/%d", i+1, len(urls))
	}
	
	fmt.Println()
	return nil
}

// GetDASHInfo 获取 DASH 信息
func (d *LegacyDASHDownloader) GetDASHInfo(mpdURL string) (map[string]interface{}, error) {
	mpd, err := d.DownloadMPD(mpdURL)
	if err != nil {
		return nil, err
	}

	baseURL := mpdURL[:strings.LastIndex(mpdURL, "/")+1]
	formats := d.ParseMPD(mpd, baseURL)

	info := map[string]interface{}{
		"type": mpd.Type,
		"duration": mpd.MediaPres,
		"formats": formats,
		"total_formats": len(formats),
	}

	return info, nil
}

// SelectQuality 选择清晰度
func (d *LegacyDASHDownloader) SelectQuality(formats []LegacyDASHFormat, quality string) *LegacyDASHFormat {
	qualityMap := map[string]int{
		"4k": 2160,
		"2k": 1440,
		"1080p": 1080,
		"720p": 720,
		"480p": 480,
		"360p": 360,
	}

	targetHeight, ok := qualityMap[strings.ToLower(quality)]
	if !ok {
		targetHeight = 1080 // 默认 1080p
	}

	var best LegacyDASHFormat
	for _, format := range formats {
		if format.Height <= targetHeight {
			if best.ID == "" || format.Height > best.Height {
				best = format
			}
		}
	}

	return &best
}
